[WIP]: Add geometric_mean aggregate function

nmahadevuni opened the pull request:
This initial implementation excludes decimal type.

nmahadevuni commented:
@aditi-pandit Can you review this?

aditi-pandit commented:
Hi @nmahadevuni : We previously spoke about consolidating this aggregate with the other SimpleNumericAggregates. DId you give that approach a try ?

nmahadevuni commented:
@aditi-pandit Please review this.

aditi-pandit commented on velox/functions/prestosql/aggregates/AverageAggregate.cpp:
```diff
@@ -134,7 +143,12 @@ class AverageAggregate : public exec::Aggregate {
         const TInput value = decodedRaw_.valueAt<TInput>(0);
         const auto numRows = rows.countSelected();
-        updateNonNullValue(group, numRows, TAccumulator(value) * numRows);
+        if (aggregateType == AggregateType::kAvg) {
```
This is a runtime conditional statement. We avoid conditionals in loops in Velox to avoid branch misses.

This code should be made a compile time check by using constexpr instead.

if constexpr (aggregateType == AggregateType::kAvg)

Please change this in all the usages in this code and then I'll take another look.

nmahadevuni commented on velox/functions/prestosql/aggregates/AverageAggregate.cpp:
```diff
@@ -359,26 +393,85 @@ bool registerAverage(const std::string& name) {
+                    int16_t,
+                    double,
+                    double,
+                    AggregateType::kAvg>>(resultType);
```
No, it won't allow variables in the template instantiation

aditi-pandit replied:
Maybe then abstract a function (which is templatized with all the types as parameters) and which does the conditional in its body so that we avoid similar logic for each type.

aditi-pandit commented on velox/functions/prestosql/aggregates/AverageAggregate.cpp:
```diff
@@ -134,7 +143,12 @@ class AverageAggregate : public exec::Aggregate {
         const TInput value = decodedRaw_.valueAt<TInput>(0);
         const auto numRows = rows.countSelected();
-        updateNonNullValue(group, numRows, TAccumulator(value) * numRows);
+        if constexpr (aggregateType == AggregateType::kAvg) {
```
It might look cleaner if value above is not const and then
if (constexpr (aggregateType == AggregateType::kGeoMean) {
  value = log10(value);
}
updateNonNullValue(...)

nmahadevuni commented:
@aditi-pandit I have addressed the comments. Please review again.

aditi-pandit commented on velox/functions/prestosql/aggregates/AverageAggregate.cpp:
```diff
@@ -398,18 +450,20 @@ exec::AggregateRegistrationResult registerAverage(const std::string& name) {
-                           .intermediateType("varbinary")
-                           .returnType("DECIMAL(a_precision, a_scale)")
-                           .build());
+  if (aggregateType == AggregateType::kAvg) {
```
Please add a comment here indicating that Decimal types are only supported for Avg and not Geometric mean.

aditi-pandit commented on velox/functions/prestosql/aggregates/AverageAggregate.cpp:
```diff
@@ -381,7 +413,27 @@ class DecimalAverageAggregate : public DecimalAggregate<TUnscaledType> {
+  return nullptr;
+}
+
+exec::AggregateRegistrationResult registerAverage(
```
You could also make registerAverage templatized on AggregateType to avoid writing getAggregateFunction. The template variable can be directly used in the make_unique<> statement.

aditi-pandit commented on velox/type/Variant.cpp:
```diff
@@ -687,6 +687,18 @@ bool variant::equalsWithEpsilon(const variant& other) const {
   if (other.isNull()) {
     return this->isNull();
   }
+  if (kind_ == TypeKind::ROW) {
```
The epsilon comparison should only happen for FLOAT/DOUBLE values within the row. Can you enhance this check to look for only FLOAT/DOUBLE columns within the row ?

nmahadevuni replied:
For ROW members, the comparison will be done appropriately based on the type. In the following lines, it checks if the type is REAL or DOUBLE and uses epsilon comparator, for the rest of the types it falls back to the equals method. Just to make the test assertion code use epsilon comparator instead of value comparator, I had to make this change.

aditi-pandit reviewed (commented):
Thanks Naveen. I haven't yet looked at the tests. Will do another pass at them.

nmahadevuni commented:
@aditi-pandit Thanks for the review. I addressed the comments. Please review. The test failure is unrelated to this PR and is happening on the main branch too.
